tests/std/tests/GH_002652_optimize_unique_ptr_destructor/test.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Test the optimization required by GH-2652, which avoids redundantly
+// checking unique_ptr's stored pointer is non-null before passing it to
+// `delete`.
+
+#include <cassert>
+#include <memory>
+
+template <int>
+struct S {
+    static bool called;
+};
+
+template <int I>
+bool S<I>::called = false;
+
+template <>
+struct std::default_delete<S<0>> {
+    using _Library_defined = default_delete; // Lies!
+
+    void operator()(S<0>* ptr) const noexcept {
+        assert(ptr == nullptr);
+        S<0>::called = true;
+    }
+};
+
+template <>
+struct std::default_delete<S<1>> {
+    void operator()(S<1>* ptr) const noexcept {
+        assert(ptr != nullptr);
+        S<1>::called = true;
+    }
+};
+
+template <>
+struct std::default_delete<S<0>[]> {

```

Comment:
Suggestion: Although this is correct, I think it's unnecessarily confusing to reuse 0 and 1 here. It would be clearer to use `S<2>[]` and `S<3>[]`, so that we can immediately see which case is being tested. (Requires several changes below, all simple.)

This would then avoid the need to reset `S<0>::called = false;` and `S<1>::called = false;`, which I consider to be a pure win.

(This is a case of the general rule to avoid reusing the same number for different meanings.)